api/v1alpha1/humiouser_types.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2020 Humio https://humio.com
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// HumioUserStateUnknown is the Unknown state of the user
+	HumioUserStateUnknown = "Unknown"
+	// HumioUserStateExists is the Exists state of the user
+	HumioUserStateExists = "Exists"
+	// HumioUserStateNotFound is the NotFound state of the user
+	HumioUserStateNotFound = "NotFound"
+	// HumioUserStateConfigError is the state of the user when user-provided specification results in configuration error, such as non-existent humio cluster
+	HumioUserStateConfigError = "ConfigError"
+)
+
+// HumioUserSpec defines the desired state of HumioUser
+type HumioUserSpec struct {
+	// ManagedClusterName refers to an object of type HumioCluster that is managed by the operator where the Humio
+	// resources should be created.
+	// This conflicts with ExternalClusterName.
+	ManagedClusterName string `json:"managedClusterName,omitempty"`
+	// ExternalClusterName refers to an object of type HumioExternalCluster where the Humio resources should be created.
+	// This conflicts with ManagedClusterName.
+	ExternalClusterName string `json:"externalClusterName,omitempty"`
+	// Username of the user in humio
+	Username string `json:"username,omitempty"`
+	// User ID of the user in humio
+	ID string `json:"id,omitempty"`
+	// FullName is the full name of the user
+	FullName string `json:"fullName,omitempty"`
+	// Email is the email of the user
+	Email string `json:"email,omitempty"`
+	// Company is the compnay of the user
+	Company string `json:"company,omitempty"`
+	// CountryCode is the compnay of the user
+	CountryCode string `json:"countryCode,omitempty"`
+	// Picture is the url to the user's profile picture
+	Picture string `json:"picture,omitempty"`
+	// IsRoot is the root setting for the user
+	IsRoot bool `json:"isRoot,omitempty"`
+	// CreatedAt is date when the user was created
```

Comment:
Do we need this as part of the spec? I can't think of a reason one would want to set this.

Reply:
I have seen people use it and it is configurable so I think it should be in the spec.

Reply:
What controls will there be around the `IsRoot` field? Could anyone push a YAML doc to the API server and give themselves root privileges?

Reply:
I agree that `CreatedAt` does seem very strange to add (at least to the Spec of the CRD). The go client involved using `Users().Add(...)` doesn't use it (same for `Users().Update(...)`. The only use of it that I see is to reflect back the `CreatedAt` timestamp for a user as for when it was created. This does not make sense in HumioUsers.Spec though. If we really want to ensure our k8s CR's store that data, then we can store it in HumioUsers.Status if we really have to, but I'm not sure we really need it. It definitely doesn't belong in the Spec though, since users cannot specify that themselves, but is given by the user upon user-creation.

Reply:
That makes sense, I'll remove it.

Reply:
> What controls will there be around the `IsRoot` field? Could anyone push a YAML doc to the API server and give themselves root privileges?

Any user that has the ability to create the resource could change the attribute yes but they could also change the cluster resource and change the authentication.